types/policy.go

```diff
@@ -170,151 +170,136 @@ func (p SpendPolicy) String() string {
 		if err != nil {
-			return SpendPolicy{}, err
+			return SpendPolicy{}, "", err
 		}
-		return PolicyAbove(n), nil
+		sp = PolicyAbove(n)
 	case "pk":
-		if len(between) != 64 {
-			return SpendPolicy{}, errors.New("invalid public key provided")
+		key := nextToken()
+		if len(key) != 64 {
+			return SpendPolicy{}, "", errors.New("hex encoded public key string length should be 64")
 		}
 		var pk PublicKey
-		if _, err := hex.Decode(pk[:], []byte(between)); err != nil {
-			return SpendPolicy{}, err
+		if _, err := hex.Decode(pk[:], []byte(key)); err != nil {
+			return SpendPolicy{}, "", err
 		}
-		return PolicyPublicKey(pk), nil
+		sp = PolicyPublicKey(pk)
 	case "thresh":
-		comma := strings.IndexByte(between, ',')
-		if comma == -1 {
-			return SpendPolicy{}, errors.New("threshold policy missing comma")
-		}
-
-		n, err := strconv.ParseUint(string(between[:comma]), 10, 8)
+		n, err := strconv.ParseUint(nextToken(), 10, 8)
 		if err != nil {
-			return SpendPolicy{}, err
-		}
-		if n == 0 {
-			return PolicyThreshold(0, nil), nil
-		}
-		afterComma := between[comma:]
-
-		bracketBegin := strings.IndexByte(afterComma, '[')
-		if bracketBegin == -1 {
-			return SpendPolicy{}, errors.New("threshold policy missing open bracket")
+			return SpendPolicy{}, "", err
 		}
-		bracketEnd := strings.LastIndexByte(afterComma, ']')
-		if bracketEnd == -1 {
-			return SpendPolicy{}, errors.New("threshold policy missing close bracket")
+		if err := consume(','); err != nil {
+			return SpendPolicy{}, "", err
 		}
-		if bracketEnd < bracketBegin {
-			return SpendPolicy{}, errors.New("threshold policy has close bracket before open bracket")
-		}
-
-		var policyStrings []string
-		var current strings.Builder
-		var openParen, closeParen int
-		policys := afterComma[bracketBegin+1 : bracketEnd]
-		for _, c := range policys {
-			// don't write leading commas
-			if !(openParen == 0 && c == ',') {
-				current.WriteRune(c)
-			}
-
-			if c == '(' {
-				openParen++
-			} else if c == ')' {
-				closeParen++
-			}
-
-			if (openParen != 0 && closeParen != 0) && (openParen == closeParen) {
-				openParen, closeParen = 0, 0
-				policyStrings = append(policyStrings, current.String())
-				current.Reset()
-			}
+		if err := consume('['); err != nil {
+			return SpendPolicy{}, "", err
 		}
 
 		var policies []SpendPolicy
-		for _, str := range policyStrings {
-			policy, err := ParseSpendPolicy(str)
+		for consume(']') != nil {
+			var policy SpendPolicy
+			policy, s, err = parseSpendPolicy(s)
 			if err != nil {
-				return SpendPolicy{}, err
+				return SpendPolicy{}, "", err
 			}
 			policies = append(policies, policy)
-		}
 
-		return PolicyThreshold(uint8(n), policies), nil
-	case "uc":
-		comma := strings.IndexByte(between, ',')
-		if comma == -1 {
-			return SpendPolicy{}, errors.New("unlock conditions policy missing comma")
+			// last policy in list will not have comma after it
+			// so don't check for error
+			consume(',')
```

Comment:
kinda janky... this will accept `thresh(1,above(2),)` as valid, won't it?

the standard solution to this problem is to add a `peek` function. Then you can write:
```go
for len(s) > 0 && peek() != ']' {
    // ... parseSpendPolicy ...
    
    if peek() != ',' {
        break
    }
    consume(',')
}
consume(']')
```